File: src/webclient/Protocol.ts
/**
 * Utility class defining the protocol used for communication between client and server.
 */
export default class Protocol {
     /** Represents a successful operation. */
     static success: string = 'success';

     /** Represents a request to delete a character. */
     static deleteChar: string = 'del';

     /** Indicates that the game has ended. */
     static GAME_ENDED: string = 'end';

     /** Represents a finished state. */
     static FINISHED: number = -1;

     /** Separator used for splitting protocol messages. */
     static SPLIT = ';';

     /** Contains various types of request messages. */
     static Request = class {
          /** Requests the text length. */
          static textLength = 'len';

          /** Requests a new text. */
          static newText = 'new';

          /** Requests the text in a multiplayer lobby. */
          static text = 'text';

          /** Requests synchronization data for all players' progress. */
          static syncData = 'sync';

          /**
           * Creates a change letter request message.
           * @param originLetter The original letter.
           * @param guessedLetter The guessed letter.
           * @returns The formatted change letter request message.
           */
          static changeLetter(
               originLetter: string,
               guessedLetter: string
          ): string {
               return `CL${Protocol.SPLIT}${originLetter}${Protocol.SPLIT}${guessedLetter}`;
          }
     };

     /** Contains various types of response messages. */
     static Response = class {
          /**
           * Splits a response message into parts using the protocol separator.
           * @param response The response message to split.
           * @returns An array of parts.
           */
          static splitResponse(response: string): string[] {
               return response.split(Protocol.SPLIT);
          }

          /**
           * Extracts the text length from a response message.
           * @param response The response message containing the text length.
           * @returns The extracted text length.
           */
          static textLength(response: string): number {
               return parseInt(response);
          }

          /**
           * Parses synchronization data from a response message.
           * @param response The response message containing synchronization data.
           * @returns An object containing game status, user IDs, and scores.
           */
          static sync(response: string): {
               gameStatus: string;
               userIds: number[];
               scores: number[];
          } {
               const parts = response.split(Protocol.SPLIT);
               const gameStatus = parts[0];
               const userIds: number[] = [];
               const scores: number[] = [];

               for (let i = 1; i < parts.length; i += 2) {
                    const userId = parts[i];
                    const score = parts[i + 1];
                    userIds.push(Number(userId));
                    scores.push(Number(score));
               }

               return { gameStatus, userIds, scores };
          }

          /** Contains various types of event messages. */
          static Event = class {
               /** Indicates an update to existing players. */
               static UpdateExistingPlayers = 'UEP';

               /** Indicates that a player has joined the game. */
               static PlayerJoined = 'PJ';

               /** Indicates that a player has left the game. */
               static PlayerLeft = 'PL';

               /** Indicates the start of a countdown. */
               static StartCountdown = 'SC';
          };
     };

     /** Contains utility methods related to sessions. */
     static Sessions = class {
          /**
           * Decrypts a username.
           * @param username The encrypted username.
           * @returns The decrypted username.
           */
          static decryptUsername(username: string): string {
               return username;
          }

          /**
           * Encrypts a username.
           * @param username The username to encrypt.
           * @returns The encrypted username.
           */
          static encryptUsername(username: string): string {
               return username;
          }
     };
}

File: backend/config/protocol.py
from typing import List


class Protocol:
    success = "success"
    DELETE_CHAR = "del"
    FINISHED: int = -1
    GAME_ENDED = "end"

    SPLIT = ";"

    class Encrypt:
        class Event:
            START_COUNTDOWN = "SC"
            UPDATE_SCORE = "US"

            @staticmethod
            def player_joined(username: str, user_id: int):
                return Protocol.Encrypt.combine("PJ", username, str(user_id))

            @staticmethod
            def player_left(user_id: int):
                return Protocol.Encrypt.combine("PJ", user_id)

            @staticmethod
            def change_letter_in_race(player_id: int, score: int) -> str:
                """change letter response.

                Args:
                    score (int): how many letters have the user guessed.

                Returns:
                    _type_: str
                """
                return Protocol.Encrypt.combine(
                    Protocol.Encrypt.Event.UPDATE_SCORE,
                    player_id,
                    score
                )

        @staticmethod
        def combine(*args):
            return Protocol.SPLIT.join(args)

        @staticmethod
        def opponents(opponents_usernames: List[str], opponents_user_ids):
            response = []
            for username, user_id in zip(opponents_usernames, opponents_user_ids):
                response.append(Protocol.Encrypt.combine(
                    username, str(user_id)))

            return Protocol.Encrypt.combine("UEP", *response)

        @staticmethod
        def text_length(text_length: int):
            return str(text_length)

        @staticmethod
        def change_letter(score: int) -> str:
            """change letter response.

            Args:
                score (int): how many letters have the user guessed.

            Returns:
                _type_: str
            """
            return str(score)

        @staticmethod
        def finished():
            return str(Protocol.FINISHED)

        @staticmethod
        def sync(user_ids: List[str], scores: List[int], game_status: str) -> str:
            """response to player's request to get the lobby's state. a response should look like:
            game_status;id1;score;id2;score;...

            * score may be Protocol.FINISHED

            Args:
                user_ids (List[str]): list of user ids
                scores (List[int]): list of scores (should be corresponding with user ids)
                game_status (str): game.status

            Returns:
                str: response
            """
            response = game_status
            for user_id, score in zip(user_ids, scores):
                response += f"{Protocol.SPLIT}{user_id}{Protocol.SPLIT}{score}"

            return response

    class Decrypt:
        @staticmethod
        def seperate_to_fields(request: str):
            return request.split(";")

        @staticmethod
        def change_letter(fields: list):
            if len(fields) != 2:
                return None
            from_letter, to_letter = fields
            return from_letter, to_letter

    class Command:
        change_letter = "CL"
        new_text = "new"

    class Request:
        text_length = "len"
        text = "text"
        sync = "sync"

    class Error:
        empty_request = "ERR.EMPTY"
        invalid_request = "ERR.INVALID"
        user_already_exists = "ERR.USER_ALREADY_EXISTS"
        not_in_game = "ERR.NOT_IN_GAME"


if __name__ == '__main__':
    print(Protocol.Encrypt.combine('hi', 'arg2', 'arg3'))
